monkey/infection_monkey/master/automated_master.py

```diff
@@ -0,0 +1,152 @@
+import logging
+import threading
+import time
+from typing import Dict, List
+
+from infection_monkey.i_control_channel import IControlChannel
+from infection_monkey.i_master import IMaster
+from infection_monkey.i_puppet import IPuppet
+from infection_monkey.telemetry.messengers.i_telemetry_messenger import ITelemetryMessenger
+from infection_monkey.telemetry.system_info_telem import SystemInfoTelem
+from infection_monkey.utils.timer import Timer
+
+CHECK_ISLAND_FOR_STOP_COMMAND_INTERVAL_SEC = 5
+CHECK_FOR_TERMINATE_INTERVAL_SEC = CHECK_ISLAND_FOR_STOP_COMMAND_INTERVAL_SEC / 5
+SHUTDOWN_TIMEOUT = 5
+
+logger = logging.getLogger()
+
+
+class AutomatedMaster(IMaster):
+    def __init__(
+        self,
+        puppet: IPuppet,
+        telemetry_messenger: ITelemetryMessenger,
+        control_channel: IControlChannel,
+    ):
+        self._puppet = puppet
+        self._telemetry_messenger = telemetry_messenger
+        self._control_channel = control_channel
+
+        self._stop = threading.Event()
+        self._master_thread = threading.Thread(target=self._run_master_thread, daemon=True)
+        self._simulation_thread = threading.Thread(target=self._run_simulation, daemon=True)
+
+    def start(self):
+        logger.info("Starting automated breach and attack simulation")
+        self._master_thread.start()
+        self._master_thread.join()
+        logger.info("The simulation has been shutdown.")
+
+    def terminate(self):
+        logger.info("Stopping automated breach and attack simulation")
+        self._stop.set()
+
+        if self._master_thread.is_alive():
+            self._master_thread.join()
```

Comment:
What is the purpose of this `join()`? There's already one right after `self._master_thread.start()`.

Reply:
Is there supposed to be a timeout here?

Reply:
There's not supposed to be a timeout. `terminate()` returns when the master thread has terminated. `terminate()` may be called by a different thread than `start()`.